modules/light-clients/xx-mock/types/update.go

```diff
@@ -1,38 +1,85 @@
+	"github.com/cosmos/ibc-go/v7/modules/core/exported"
 )
 
-// CheckHeaderAndUpdateState checks if the provided header is valid and updates
-// the consensus state if appropriate. It returns an error if:
-// - the header provided is not parseable to a Mock header
-func (cs ClientState) CheckHeaderAndUpdateState(
+// VerifyClientMessage checks if the clientMessage is of type Header
+func (cs *ClientState) VerifyClientMessage(
 	ctx sdk.Context, cdc codec.BinaryCodec, clientStore sdk.KVStore,
-	header exported.Header,
-) (exported.ClientState, exported.ConsensusState, error) {
-	smHeader, ok := header.(*Header)
-	if !ok {
-		return nil, nil, sdkerrors.Wrapf(
-			clienttypes.ErrInvalidHeader, "header type %T, expected  %T", header, &Header{},
-		)
+	clientMsg exported.ClientMessage,
+) error {
+	switch msg := clientMsg.(type) {
+	case *Header:
+		return cs.verifyHeader(ctx, clientStore, cdc, msg)
+	default:
+		return clienttypes.ErrInvalidClientType
 	}
+}
+
+// verifyHeader returns an error if:
+// - header revision is not equal to latest header revision
+func (cs *ClientState) verifyHeader(
+	ctx sdk.Context, clientStore sdk.KVStore, cdc codec.BinaryCodec,
+	header *Header,
+) error {
+	if header.GetHeight().GetRevisionNumber() != cs.LatestHeight.RevisionNumber {
+		return sdkerrors.Wrapf(
+			ErrInvalidHeaderHeight,
+			"header height revision %d does not match latest header revision %d",
+			header.GetHeight().GetRevisionNumber(), cs.LatestHeight.RevisionNumber,
+		)
 
-	clientState, consensusState := update(&cs, smHeader)
-	return clientState, consensusState, nil
+	}
+	return nil
 }
 
-// update the consensus state to the new public key and an incremented revision number
-func update(clientState *ClientState, header *Header) (*ClientState, *ConsensusState) {
+// UpdateState may be used to either create a consensus state for:
+// - a future height greater than the latest client state height
+// - a past height that was skipped during bisection
+// If we are updating to a past height, a consensus state is created for that height to be persisted in client store
+// If we are updating to a future height, the consensus state is created and the client state is updated to reflect
+// the new latest height
+// A list containing the updated consensus height is returned.
+// UpdateState must only be used to update within a single revision, thus header revision number and trusted height's revision
+// number must be the same. To update to a new revision, use a separate upgrade path
+func (cs ClientState) UpdateState(ctx sdk.Context, cdc codec.BinaryCodec, clientStore sdk.KVStore, clientMsg exported.ClientMessage) []exported.Height {
+	header, ok := clientMsg.(*Header)
+	if !ok {
+		panic(fmt.Errorf("expected type %T, got %T", &Header{}, clientMsg))
+	}
+
+	// check for duplicate update
+	if _, found := getConsensusState(clientStore, cdc, header.GetHeight()); found {
+		// perform no-op
+		return []exported.Height{header.GetHeight()}
```

Comment:
should we return an empty slice instead here?

Reply:
I haven't been able to find the exact spec on this, but Tendermint client returns one height in the same case.
https://github.com/cosmos/ibc-go/blob/v7.0.0/modules/light-clients/07-tendermint/update.go#L141-L142

Reply:
Yes, it is true, but I think it does not satisfy the following comment at least:
https://github.com/cosmos/ibc-go/blob/v7.0.0/modules/light-clients/07-tendermint/update.go#LL127C4-L127C63
> A list containing the updated consensus height is returned.

Reply:
I see. I will ask about this in the ibc-go repo.

Reply:
I found this by going back through the conversation history on GitHub.
https://github.com/cosmos/ibc-go/pull/1285/commits/9116fd127152e6fee2cc3a19c9807a6788e6db0e#r857602843

Reply:
To sum up, heights returned by `UpdateState` are used for logging and events.
If the slice is empty at the time, there may be a problem.

Reply:
@bluele As a conclusion, I'm thinking of keeping it as is.